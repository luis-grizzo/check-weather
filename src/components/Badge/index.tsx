interface BadgeProps {
  children: string
}

export function Badge({ children }: BadgeProps) {
  return (
    <span className="flex items-center gap-2 px-2 py-1 text-xs bg-neutral-100/60 border-1 border-neutral-600/10 text-neutral-950 rounded-lg backdrop-blur">
      {children}
    </span>
  )
}
