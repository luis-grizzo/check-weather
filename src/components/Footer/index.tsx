'use client'

export function Footer() {
  const currentYear = new Date().getFullYear()

  return (
    <footer className="bg-neutral-100/60 border-t-1 border-neutral-600/10">
      <p className="container mx-auto px-8 py-6 text-xs text-center">
        Developed by{' '}
        <a
          href="https://luisgrizzo.dev"
          target="_blank"
          rel="noreferrer"
          className="text-blue-600 hover:text-blue-400 transition-colors"
        >
          Lu√≠s Grizzo
        </a>
        {', '}
        <a
          href="https://github.com/luis-grizzo/check-weather?tab=MIT-1-ov-file"
          target="_blank"
          className="text-xs text-blue-600 hover:text-blue-400 transition-colors"
          rel="noreferrer"
        >
          MIT License
        </a>
        {', '}
        {currentYear}.
      </p>
    </footer>
  )
}
