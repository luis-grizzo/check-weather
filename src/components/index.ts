export * from './Badge'
export * from './Button'
export * from './DisplayTime'
export * from './Footer'
export * from './Header'
export * from './InfoPanel'
export * from './Navbar'
export * from './RefreshButton'
