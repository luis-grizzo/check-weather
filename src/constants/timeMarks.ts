export const hourInMilliseconds = 3_600_000

export const hourInSeconds = 3_600
