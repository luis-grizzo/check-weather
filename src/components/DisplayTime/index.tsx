'use client'

import { Badge } from '@/components/Badge'

import { formatTime } from '@/lib/stringFormaters'

interface DisplayTimeProps {
  description: string
  timestamp: number
}

export function DisplayTime({ description, timestamp }: DisplayTimeProps) {
  return <Badge>{`${description} ${formatTime(timestamp)}`}</Badge>
}
