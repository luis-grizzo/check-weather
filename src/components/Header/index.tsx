'use client'

import { useEffect, useState } from 'react'

import { hourInMilliseconds } from '@/constants/timeMarks'

import {
  formatCountryName,
  formatDate,
  formatTime,
  formatTimer
} from '@/lib/stringFormaters'
import { positiveOrZero } from '@/lib/numberFormaters'

import type { FetchWeatherFactoredResponse } from '@/utils/weatherUtils'
import { useRouter } from 'next/navigation'

interface HeaderProps {
  location: FetchWeatherFactoredResponse['location']
  requestTimestamp: number
}

export function Header({ location, requestTimestamp }: HeaderProps) {
  const router = useRouter()

  const [countdownTimer, setCountdownTimer] = useState('--:--')
  const [shouldRefresh, setShouldRefresh] = useState(false)

  useEffect(() => {
    const timer = setInterval(() => {
      const currentTime = Date.now()
      const revalidateTime = requestTimestamp * 1_000 + hourInMilliseconds

      const remainTime = revalidateTime - currentTime
      const validatedRemainTime = positiveOrZero(remainTime)

      const formattedTimer = formatTimer(validatedRemainTime)

      setCountdownTimer(formattedTimer)

      if (!validatedRemainTime) {
        setShouldRefresh(true)

        clearInterval(timer)
      }
    }, 1_000)

    return () => clearInterval(timer)
  }, [requestTimestamp])

  useEffect(() => {
    if (shouldRefresh) {
      router.refresh()

      setShouldRefresh(false)
    }
  }, [shouldRefresh])

  return (
    <header className="flex items-center justify-between container mx-auto px-8 py-6">
      <div className="flex flex-col gap-1">
        <span role="timer" className="text-xs">
          {`Next forecast in ${countdownTimer}`}
        </span>

        {location.country && (
          <span className="text-lg">
            {`${location.city}, ${formatCountryName(location.country)}`}
          </span>
        )}

        <span className="text-xs">{`${formatTime(requestTimestamp)} - ${formatDate(requestTimestamp)}`}</span>
      </div>
    </header>
  )
}
