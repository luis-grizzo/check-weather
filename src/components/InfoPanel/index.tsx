import Image from 'next/image'

import { layoutIcons } from '@/constants/icons'

interface InfoPanelProps {
  children: React.ReactNode
}

export function InfoPanel({ children }: InfoPanelProps) {
  return (
    <div
      role="dialog"
      className="flex items-start gap-4 p-4 text-sm text-red-950 bg-red-100/60 border-red-600/10 border-1 rounded-xl"
    >
      <Image src={layoutIcons.error} alt="" className="w-5 aspect-square" />

      {children}
    </div>
  )
}
