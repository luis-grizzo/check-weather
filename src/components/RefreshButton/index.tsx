'use client'

import { useEffect, useState } from 'react'
import { useRouter } from 'next/navigation'
import Image from 'next/image'

import { layoutIcons } from '@/constants/icons'
import { hourInMilliseconds } from '@/constants/timeMarks'
import { formatTimer } from '@/lib/stringFormaters'

interface RefreshButtonProps {
  unixTimestampOfLastRequest: number
}

export function RefreshButton({
  unixTimestampOfLastRequest
}: RefreshButtonProps) {
  const router = useRouter()

  const [isDisabled, setIsDisabled] = useState(true)
  const [countdownTimer, setCountdownTimer] = useState<string | undefined>(
    undefined
  )

  const refresh = () => router.refresh()

  useEffect(() => {
    const timer = setInterval(() => {
      const currentTime = Date.now()
      const revalidateTime =
        unixTimestampOfLastRequest * 1_000 + hourInMilliseconds

      const remainTime = revalidateTime - currentTime
      const formattedTimer = formatTimer(remainTime)

      setCountdownTimer(formattedTimer)

      setIsDisabled(!!remainTime)
    }, 1_000)

    if (!isDisabled) clearInterval(timer)

    return () => clearInterval(timer)
  }, [unixTimestampOfLastRequest, isDisabled])

  return (
    <div className="flex flex-col items-center gap-2">
      <button
        className="flex items-center justify-center p-2 bg-neutral-100/60 border-1 border-neutral-600/10 rounded-full disabled:opacity-50 disabled:cursor-not-allowed transition-opacity"
        onClick={refresh}
        disabled={isDisabled}
      >
        <Image
          src={
            countdownTimer ? layoutIcons.refresh : layoutIcons.progress_activity
          }
          alt={countdownTimer ? 'Refresh icon' : 'Loading icon'}
          className={`w-6 aspect-square ${!countdownTimer ? 'animate-spin' : ''}`}
        />
      </button>

      {countdownTimer && <span className="text-xs">{countdownTimer}</span>}
    </div>
  )
}
