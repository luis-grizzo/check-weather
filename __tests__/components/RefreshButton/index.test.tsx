import { act } from 'react'
import { fireEvent, render, screen, waitFor } from '@testing-library/react'

import { hourInMilliseconds } from '@/constants/timeMarks'

import { RefreshButton } from '@/components/RefreshButton'

const unixTimestamp = 1718729676
const secondInMilliseconds = 1_000

jest.useFakeTimers()
jest.setSystemTime(
  new Date(unixTimestamp * 1_000 + hourInMilliseconds - secondInMilliseconds)
)

const refreshMock = jest.fn()

jest.mock('next/navigation', () => ({
  useRouter() {
    return {
      refresh: refreshMock
    }
  }
}))

describe('RefreshButton', () => {
  afterAll(() => {
    jest.clearAllMocks()
    jest.clearAllTimers()
  })

  beforeEach(() => {
    render(<RefreshButton unixTimestampOfLastRequest={unixTimestamp} />)
  })

  it('should match snapshot', () => {
    const { container } = render(
      <RefreshButton unixTimestampOfLastRequest={unixTimestamp} />
    )

    expect(container).toMatchSnapshot()
  })

  it('should enable button when the timer reaches zero', async () => {
    const eButton = screen.getByRole('button') as HTMLButtonElement

    expect(eButton.disabled).toBeTruthy()

    act(() => {
      jest.advanceTimersByTime(2_000)
    })

    await waitFor(() => {
      const eButton = screen.getByRole('button') as HTMLButtonElement

      expect(eButton.disabled).toBeFalsy()
    })
  })

  it('should call router refresh on button click', async () => {
    act(() => {
      jest.advanceTimersByTime(2_000)
    })

    await waitFor(() => {
      const eButton = screen.getByRole('button') as HTMLButtonElement

      fireEvent.click(eButton)

      expect(refreshMock).toHaveBeenCalled()
    })
  })
})
