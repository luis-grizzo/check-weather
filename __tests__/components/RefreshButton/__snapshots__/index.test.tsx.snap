// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RefreshButton should match snapshot 1`] = `
<div>
  <div
    class="flex flex-col items-center gap-2"
  >
    <button
      class="flex items-center justify-center p-2 bg-neutral-100/60 border-1 border-neutral-600/10 rounded-full disabled:opacity-50 disabled:cursor-not-allowed transition-opacity"
      disabled=""
      role="button"
    >
      <img
        alt="Loading icon"
        class="w-6 aspect-square animate-spin"
        data-nimg="1"
        decoding="async"
        height="40"
        loading="lazy"
        src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
        srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
        style="color: transparent;"
        width="40"
      />
    </button>
  </div>
</div>
`;
